src/core/tsi/ssl_transport_security.cc

```diff
@@ -853,6 +882,38 @@ static int NullVerifyCallback(int /*preverify_ok*/, X509_STORE_CTX* /*ctx*/) {
   return 1;
 }
 
+static int RootCertExtractCallback(int preverify_ok, X509_STORE_CTX* ctx) {
+  // There's a case where this function is set in SSL_CTX_set_verify and a CRL
+  // related callback is set with X509_STORE_set_verify_cb. They overlap and
+  // this will take precedence, thus we need to ensure the CRL related callback
+  // is still called
+  X509_VERIFY_PARAM* param = X509_STORE_CTX_get0_param(ctx);
+  auto flags = X509_VERIFY_PARAM_get_flags(param);
+  if (flags & X509_V_FLAG_CRL_CHECK) {
+    preverify_ok = verify_cb(preverify_ok, ctx);
+  }
+
+  if (preverify_ok == 0) {
+    return 0;
+  }
+
+  // If we're here, verification was successful
+  // Get the verified chain from the X509_STORE_CTX and put it on the SSL object
+  // so that we have access to it when populating the tsi_peer
+  STACK_OF(X509)* chain = X509_STORE_CTX_get0_chain(ctx);
+  if (chain == nullptr) {
+    return 1;
+  }
+
+  // The ca cert is the last in the chain
+  X509* ca_cert = sk_X509_value(chain, sk_X509_num(chain) - 1);
```

Comment:
Please return early if `ca_cert` is nullptr.

Reply:
Ack, will change